### PR TITLE
feat(web): Caps Layer and double-tap gesture 🕊

### LINKED ISSUE #3620: spec: Caps Lock layer for touch layouts
This is part of a broader specification for Caps Lock support on touch devices. Development of this touches multiple components; each component will be implemented in a separate PR, referencing this issue.

## Related features

* #3621: spec: Start of text/sentence selects shift layer
* #3720: spec: Suggestions respect capitalization
* #5029: spec: multitap and flick for touch keyboards

## Introduction

A separate `caps` layer should be supported for touch layouts that provides support for Caps Lock semantics. This allows for slight nuances if there are differences between a `caps` and a `shift` layer, and also gives a clean pathway forward for existing keyboards that may have already defined `nextlayer` properties on each key.

If the keyboard defines layer controls in rules, these will not override `nextlayer` controls. According to [spec](https://help.keyman.com/developer/language/reference/layer), touch layout wins.

The `caps` layer, if present, will be accessible by double-tapping the Shift key. See #5029.

## C1.1 File Formats: .keyman-touch-layout changes

Changes to the file format and the JSON schema are required.

1. Required: Add a `caps` entry to the list of predefined layers. The templates will not include this layer by default, as ideally Keyman Developer will include tooling to help generate the `caps` layer from the `shift` layer.
2. Optional: Add a property for each layer called `nextlayer`. When set, for each normal key on that layer that has no `nextlayer` set, this would be applied by KeymanWeb. (this resolves the tedium as discussed here.) The templates would be updated to set `nextlayer` to `"default"` for `shift` layers.

###  Backward compatibility

Touch layout files that contain these features will be backwardly compatible with earlier versions of Keyman, with a graceful degradation of functionality.

* (if `nextlayer` is implemented) The compiler will transform the `nextlayer` property from layer-level to key-level at build time. Existing per-key `nextlayer` properties would override the layer `nextlayer` property.

## C1.2 Changes to Compiler

1. (if implemented) The compiler will transform the `nextlayer` property from layer-level to key-level at build time. Existing per-key `nextlayer` properties would override the layer `nextlayer` property.

## C1.3 Changes to Developer IDE

* Optional: The touch layout editor needs to surface the `nextlayer` property for each layer.
* Optional: A helpful function would be to be able to generate the `caps` layer from the IDE.
  * A caps layer’s initial form may easily be made by duplicating its base layer.
  * No special wizard or anything needed there.
  * Keys that have overridden modifiers may need manual correction.
  * nextlayer properties on each key may need manual correction.
  * Updates to the shift layer will not auto-propagate to the caps layer.
* Optional: Also helpful would be some validation tests for `nextlayer` property consistency.

## C1.4 KeymanWeb enhancements

* Required: KeymanWeb needs to support the Caps Lock layer.

**Caps Lock Interaction:** double-tapping Shift will switch to the `caps` layer -- see #5029.

**Caps Lock Presentation:** the `caps` layer Shift key needs treatment to differentiate from the Shift key on the `shift` layer. This will be through the caps indicator image defined as `*Caps*`. We should also consider colour styling, for example, this could be defined in the CSS with a rule for selector `.kmw-key-shift#caps-K_SHIFT`. The styling may differ per platform.

## Related Issues and discussions

* #246
* https://community.software.sil.org/t/how-to-return-from-shift-automatically-and-how-to-caps-lock-by-double-tapping/3469/3 
* https://community.software.sil.org/t/problem-with-caps-lock-output/3718
* [Original design document](https://docs.google.com/document/d/1J1WX41trb7GtPMM_VkfnV3XNOd3WFfHRlXdbf6YR7U8/edit#heading=h.vn6hrktkdivc)
* [RFC](https://community.software.sil.org/t/request-for-comment-specification-for-caps-lock-layer-for-touch-layouts/3890/2)
* [15.0 planning document](https://docs.google.com/document/d/1ojMK69QMFQzuZFfyGHphRO4eBiPJvx_nOoDndabVXsY/edit#)

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #3620.

Implements the Caps Lock layer support and the double-tap gesture on the shift key to access it.

The double-tap gesture has been implemented with a view to extension to support other multi-tap gestures in the future. However, for now, it is limited to supporting the Shift key, if and only if the keyboard includes a Caps layer.

The reason for this v15 limitation is that multi-tap on regular keys would involve either rewinding the previous keystroke (the first tap), or forcing keyboard developers to consider 'rota' style rules in their keyboards to support the multi-tap gestures, as we need to make sure that the first tap is accepted and processed for immediate feedback. This needs more design, to avoid unnecessary complexity in the keyboards and/or the rewinding of the keystroke (even though that is conceptually supported in Keyman Engine for Web already). Basically, we don't want to constrain the way that a keyboard author may use the multi-tap gesture by hard-coding the rewind, but neither do we want to make all multi-tap gestures needlessly complex to author.

The shift key (and other modifiers, potentially in future) needs special support for multi-tap as the key that is being tapped changes with the layer change. This is currently managed through recognising `K_SHIFT` in the key id.

I have tried to follow the `PendingGesture` pattern for multi-tap, and the gesture itself supports a series of taps, not just a double-tap. The maximum time to complete the tap series is 125msec * number-of-taps, so for a double-tap is 250msec.

The changes to support a Caps Lock layer itself were minimal; just adding the `text.KeyboardProcessor.getStateFromLayer` function and calling it during `KeyEvent` construction. The remaining changes relate to the multi-tap gesture.

Minor changes:
* I moved `constructNullKeyEvent` to `KeyEvent` in order to make it more accessible to other classes.
* The multi-tap gesture does not have a promise to complete, so that is now an optional member of the `PendingGesture` interface.

Note:
* The testing page for KeymanWeb includes a test for just the Caps Lock layer support, caps_lock_layer_3620.js, and a full Caps Lock and Start of Sentence test, full_caps_3620_3621.js.

# User Testing

### SUITE_BASELINE: Test that nothing has changed in existing keyboards

1. Run through basic usage tests of a variety of keyboards, including sil_euro_latin and khmer_angkor.
2. Test longpress keys, layer switching, flick-up gesture.
3. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

* **TEST_BASELINE_WEB:** On testing/unminified, run through the above tests. Make sure no script errors are encountered. Do this both for desktop and touch modes.
* **TEST_BASELINE_ANDROID:** In Keyman for Android, on an Android device or emulator, run through the above tests, both system and in-app keyboards.
* **TEST_BASELINE_IPHONE:** In Keyman for iPhone, on an iPhone device or emulator, run through the above tests, both system and in-app keyboards.
* **TEST_BASELINE_IPAD:** In Keyman for iPad, on an iPad device or emulator, run through the above tests, both system and in-app keyboards.

### SUITE_CAPS: Test that the new Caps Lock layer controls work correctly

1. Run through a basic usage test of [caps_lock_layer_3620.zip](https://github.com/keymanapp/keyman/files/7731689/caps_lock_layer_3620.zip).
2. Double-tap the shift key. Watch how it switches to a Caps layer (see the 'shiftlock' style of the Shift key).
3. Test that the Caps layer works correctly. Note that the Shift layer on caps_lock_layer_3620 does not shift back to the base layer automatically; this is intentional.
4. With the [full_caps_3620_3621.zip](https://github.com/keymanapp/keyman/files/7731695/full_caps_3620_3621.zip) keyboard, test that the Caps layer is accessible through double-tap shift, and that the interactions with Caps Lock and Shift are 'intuitive'.
5. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

* **TEST_CAPS_WEB:** On testing/"Test Caps Lock Layer (#3620)", run through the above tests. Make sure no script errors are encountered.
* **TEST_CAPS_ANDROID:** In Keyman for Android, on an Android device if possible, run through the above tests, both system and in-app keyboards.
* **TEST_CAPS_IPHONE:** In Keyman for iPhone, on an iPhone device if possible, run through the above tests, both system and in-app keyboards.
* **TEST_CAPS_IPAD:** In Keyman for iPad, on an iPad device if possible, run through the above tests, both system and in-app keyboards.

### CODE REVIEW

File: common/core/web/keyboard-processor/src/text/keyboardProcessor.ts [hunk 1]
```diff
@@ -367,6 +364,22 @@ namespace com.keyman.text {
       return modifier;
     }
 
+    /**
+     * Get state key state from layer id
```
Comment (ermshiperete) [inline common/core/web/keyboard-processor/src/text/keyboardProcessor.ts:368]: Is the double `state` intentional?

File: common/test/keyboards/full_caps_3620_3621/source/full_caps_3620_3621.kmn [hunk 2]
```diff
@@ -25,11 +25,10 @@ group(NewContext) readonly
...
 group(PostKeystroke) readonly
-    c this is a heuristic that we can use until we support a CAPS layer (#3620):
-    c if we find two+ characters in capitals, leave us in shift layer,
-    c assuming the user hasn't just switched layer themselves. Once we have a
-    c proper CAPS layer, we'll get rid of this.
-    if(&layerChanged = "0") any(caps) any(caps) > context
+    c this is a heuristic: if we find two+ characters in capitals, leave
+    c us in shift layer, assuming the user hasn't just switched layer
+    c themselves. Not sure if this is a good idea!
+    c if(&layerChanged = "0") if(&layer = 'shift') any(caps) any(caps) > layer('caps') c ??? is this a good idea?
```
Comment (ermshiperete) [inline common/test/keyboards/full_caps_3620_3621/source/full_caps_3620_3621.kmn:31]: do you want to keep this comment?

File: web/testing/caps-lock-layer-3620/caps_lock_layer_3620.js [hunk 3]
```diff
@@ -0,0 +1,1661 @@
...
+  var keyCodes = keyman.osk.keyCodes;
+
+  this._v=(typeof keyman!="undefined"&&typeof keyman.version=="string")?parseInt(keyman.version,10):9;
+  this.KI="Keyboard_caps_lock_layer_3620";
+  this.KN="Caps Lock Layer 3620";
+  this.KMINVER="10.0";
+  this.KV={F:' 1em "Arial"',K102:0};
+  this.KV.KLS={
+
+  };
```
Comment (ermshiperete) [inline web/testing/caps-lock-layer-3620/caps_lock_layer_3620.js:19]:
```suggestion
  this.KV.KLS={ };
```

**Review by ermshiperete — APPROVED** [review]
Didn't see anything obvious, but my knowledge in that part of the code is pretty limited :smile:

File: common/core/web/keyboard-processor/src/text/keyboardProcessor.ts (hunk 1 above)
Comment (mcdurdin) [reply common/core/web/keyboard-processor/src/text/keyboardProcessor.ts:368]: Yes, on 101 key keyboards, the 'state' keys are the caps lock, num lock, scroll lock keys, but we only care about Caps Lock. So this is retrieving the state of these keys 😆.

Both `getStateFromLayer` and `getModifierState` above could be better; right now a layer called "foolscapsymbols" (okay, it's a stretch) would also be considered a 'caps lock' layer.

File: web/testing/caps-lock-layer-3620/caps_lock_layer_3620.js (hunk 3 above)
Comment (mcdurdin) [reply web/testing/caps-lock-layer-3620/caps_lock_layer_3620.js:19]: This is compiled code; only reason I've included it in the PR is to make it available for testers.

File: common/test/keyboards/full_caps_3620_3621/source/full_caps_3620_3621.kmn (hunk 2 above)
Comment (mcdurdin) [reply common/test/keyboards/full_caps_3620_3621/source/full_caps_3620_3621.kmn:31]: Perhaps I should probably remove that heuristic altogether? I will leave it for now but think about it for a follow-up PR.

**Comment by MakaraSok** [comment]
## SUITE_BASELINE: Test that nothing has changed in existing keyboards
- **TEST_BASELINE_WEB: PASSED** The OSK seems to work fine, but there is a catch. I don't know what "script error" would entail, but as far as the usage of the OSK is concerned, nothing stops me from using the keyboards.

  The OSK doesn't show keycaps in capital letters when <kbd>Caps</kbd> key is active.
  
  

  On Touch devices, everything works without any issue, but on iPad: the characters on keycaps are kinda a little too small.

    

    

    

- **TEST_BASELINE_ANDROID: FAILED?** 
  * In-app: longpress keys are not working in simulator, but OK in a physical device.

    https://user-images.githubusercontent.com/28331388/145956348-a13b1f93-c616-4639-9a9b-4ec73f30c762.mov
  
  * System: no obvious issue found, but a glitch found in the rapid touch where the preview key stops showing after some time. 

    https://keymanapp.slack.com/files/U6Q7Y6XTM/F02QMS0KDQT/screen_recording_20211214-155720_keyman.mp4
  
- **TEST_BASELINE_IPHONE: PASSED?** 
  * In-app: No obvious issue found, but there is always a keypress/tap noice when the keyboard finishes starting up. I've never heard this before. Two keypresses/taps are heard in-app and one as a system keyboard. Keyman sometimes crashes (see the beginning of the recording below.)

    https://user-images.githubusercontent.com/28331388/145947540-d938cc36-7d93-4adf-b9ba-78a5db362704.mov

  Note also that the "Get Started" page doesn't show up even though it's on in the settings.
  
  * System: no issue other than the mysterious keypress/tap noice.
  
- **TEST_BASELINE_IPAD: PASSED** No obvious error is found.
  * In-app: Unlike on iPhone, the "Get Started" page is always shown at start up. I've noticed that one keypress/tap noice is heard before the page is shown and another after the page is closed.

    https://user-images.githubusercontent.com/28331388/145950764-baf49665-f570-45ca-8ee2-bff0a098c524.mov
  
  * System: no issue other than the mysterious keypress/tap noice.

**Comment by mcdurdin** [comment]
> * Make sure no script errors are encountered

To check for script errors when testing KeymanWeb, open the Developer Console (F12 on Chrome/Windows) -- script errors are shown in red and can be located by clicking the 
image button in the top right of the Developer Console.

**Comment by mcdurdin** [comment]
> * System: no obvious issue found, but a glitch found in the rapid touch where the preview key stops showing after some time.
>   https://keymanapp.slack.com/files/U6Q7Y6XTM/F02QMS0KDQT/screen_recording_20211214-155720_keyman.mp4

This video no longer appears available?

**Comment by mcdurdin** [comment]
> * [TEST_BASELINE_ANDROID] In-app: longpress keys are not working in simulator, but OK in a physical device.

This may not be related to this change; can we do a screenshare test together on this?

@keymanapp-test-bot retest SUITE_BASELINE TEST_BASELINE_ANDROID

**Comment by MakaraSok** [comment]
> > * System: no obvious issue found, but a glitch found in the rapid touch where the preview key stops showing after some time.
> >   https://keymanapp.slack.com/files/U6Q7Y6XTM/F02QMS0KDQT/screen_recording_20211214-155720_keyman.mp4
> 
> This video no longer appears available?

https://user-images.githubusercontent.com/28331388/146703733-ae7598e8-b306-4ace-871a-f76fd3df7d64.mp4

**Comment by MakaraSok** [comment]
### SUITE_CAPS: Test that the new Caps Lock layer controls work correctly

- **TEST_BASELINE_ANDROID: FAILED??**

1. Run through a basic usage test of [caps_lock_layer_3620.zip](https://github.com/keymanapp/keyman/files/7731689/caps_lock_layer_3620.zip).

> One of the basic usage of the keyboard doesn't work as expected. Long press keys are not responsive; the subkeys do not pop up as expected on Android 10 simulator API 29. Note that this doesn't happen on the physical device.

2. Double-tap the shift key. Watch how it switches to a Caps layer (see the 'shiftlock' style of the Shift key).

> Working OK.

3. Test that the Caps layer works correctly. Note that the Shift layer on caps_lock_layer_3620 does not shift back to the base layer automatically; this is intentional. 

> Each key is working OK, but those keys with longpresses on the simulator. On the physical device however, this doesn't happen.

4. With the [full_caps_3620_3621.zip](https://github.com/keymanapp/keyman/files/7731695/full_caps_3620_3621.zip) keyboard, test that the Caps layer is accessible through double-tap shift, and that the interactions with Caps Lock and Shift are 'intuitive'.

> Each key is working OK, but those keys with longpresses. On the physical device however, this doesn't happen.

5. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

> See the screen recording above.

**Comment by MakaraSok** [comment]
### SUITE_BASELINE: Test that nothing has changed in existing keyboards

* **TEST_BASELINE_ANDROID:** In Keyman for Android, on an Android device or emulator, run through the above tests, both system and in-app keyboards.

1. Run through basic usage tests of a variety of keyboards, including sil_euro_latin and khmer_angkor.

> OK, but longpress doesn't work on emulator. Nothing like this is seen on a physical device.

2. Test longpress keys, layer switching, flick-up gesture.

> OK

3. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

> The key preview may disable during a speed tapping. See the screenshot recording in the comment right above.

**Comment by MakaraSok** [comment]
### SUITE_CAPS: Test that the new Caps Lock layer controls work correctly

- **TEST_CAPS_ANDROID: FAILED??**

1. Run through a basic usage test of [caps_lock_layer_3620.zip](https://github.com/keymanapp/keyman/files/7731689/caps_lock_layer_3620.zip).

> One of the basic usage of the keyboard doesn't work as expected. Long press keys are not responsive; the subkeys do not pop up as expected on Android 10 simulator API 29. Note that this doesn't happen on the physical device.

2. Double-tap the shift key. Watch how it switches to a Caps layer (see the 'shiftlock' style of the Shift key).

> Working OK.

3. Test that the Caps layer works correctly. Note that the Shift layer on caps_lock_layer_3620 does not shift back to the base layer automatically; this is intentional. 

> Each key is working OK, but those keys with longpresses on the simulator. On the physical device however, this doesn't happen.

4. With the [full_caps_3620_3621.zip](https://github.com/keymanapp/keyman/files/7731695/full_caps_3620_3621.zip) keyboard, test that the Caps layer is accessible through double-tap shift, and that the interactions with Caps Lock and Shift are 'intuitive'.

> Each key is working OK, but those keys with longpresses. On the physical device however, this doesn't happen.

5. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

> The test is not quite practical, but it does appear that the preview went off and on amid typing.

https://user-images.githubusercontent.com/28331388/146709335-7343e4c0-dc42-419d-b042-48551177f31a.mp4

**Comment by MakaraSok** [comment]
### SUITE_CAPS: Test that the new Caps Lock layer controls work correctly

- **TEST_CAPS_IPHONE: FAILED**

1. Run through a basic usage test of [caps_lock_layer_3620.zip](https://github.com/keymanapp/keyman/files/7731689/caps_lock_layer_3620.zip).

> Nothing weird observed.

2. Double-tap the shift key. Watch how it switches to a Caps layer (see the 'shiftlock' style of the Shift key).

> Working OK, but the weird thing is the double tap seem unnatural to the touch. You don't hear the two taps, just one is heard. :/

3. Test that the Caps layer works correctly. Note that the Shift layer on caps_lock_layer_3620 does not shift back to the base layer automatically; this is intentional. 

> Each key is working OK including longpresses.

4. With the [full_caps_3620_3621.zip](https://github.com/keymanapp/keyman/files/7731695/full_caps_3620_3621.zip) keyboard, test that the Caps layer is accessible through double-tap shift, and that the interactions with Caps Lock and Shift are 'intuitive'.

> Not only that the double tap is unnatural to the touch, it doesn't enable caps layer at all. Double tap the Shift key and start typing SS gives you Ss instead.

5. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

> No physical device within reach right now.

**Comment by MakaraSok** [comment]
### SUITE_CAPS: Test that the new Caps Lock layer controls work correctly

- **TEST_CAPS_IPAD: FAILED**

1. Run through a basic usage test of [caps_lock_layer_3620.zip](https://github.com/keymanapp/keyman/files/7731689/caps_lock_layer_3620.zip).

> Nothing weird observed.

2. Double-tap the shift key. Watch how it switches to a Caps layer (see the 'shiftlock' style of the Shift key).

> Working OK, but the weird thing is the double tap seem unnatural to the touch. You don't hear the two taps, just one is heard. :/

3. Test that the Caps layer works correctly. Note that the Shift layer on caps_lock_layer_3620 does not shift back to the base layer automatically; this is intentional. 

> Each key is working OK including longpresses.

4. With the [full_caps_3620_3621.zip](https://github.com/keymanapp/keyman/files/7731695/full_caps_3620_3621.zip) keyboard, test that the Caps layer is accessible through double-tap shift, and that the interactions with Caps Lock and Shift are 'intuitive'.

> Not only that the double tap is unnatural to the touch, it doesn't enable caps layer at all. Double tap the Shift key and start typing SS gives you Ss instead.

https://user-images.githubusercontent.com/28331388/146721426-8d558d95-101d-461f-9477-322b12a1e7b0.mov


5. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

> No physical device within reach right now.

**Comment by MakaraSok** [comment]
### SUITE_CAPS: Test that the new Caps Lock layer controls work correctly

- **TEST_CAPS_WEB: PASSED**

1. Run through a basic usage test of [caps_lock_layer_3620.zip](https://github.com/keymanapp/keyman/files/7731689/caps_lock_layer_3620.zip).

> Nothing weird observed.

2. Double-tap the shift key. Watch how it switches to a Caps layer (see the 'shiftlock' style of the Shift key).

> Working OK.

3. Test that the Caps layer works correctly. Note that the Shift layer on caps_lock_layer_3620 does not shift back to the base layer automatically; this is intentional. 

> Each key is working OK including longpresses.

4. With the [full_caps_3620_3621.zip](https://github.com/keymanapp/keyman/files/7731695/full_caps_3620_3621.zip) keyboard, test that the Caps layer is accessible through double-tap shift, and that the interactions with Caps Lock and Shift are 'intuitive'.

> OK


5. For testing on a physical device: test that rapid touches across multiple keys works correctly, as does multi-touch (where you start tapping the next key before releasing the previous, e.g. rapid two-thumb typing). These are not easy to test in an emulator!

> No physical device within reach right now.

**Comment by mcdurdin** [comment]
Failing test feedback:

- Android: the longpress issue on emulator is captured at #6113.
- iOS: the issue with in-app context loss causing the caps lock layer to cancel is probably the same as #6112. 
- iOS: the issue with double-tap not giving a double click is now captured in #6121.

As I have now captured the three failure cases as separate issues, I will go ahead and merge this PR, with the caveat that these issues must be resolved before release.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.180-alpha](https://keyman.com/downloads/releases/alpha/15.0.180)

**Comment by MayuraVerma** [comment]
How to enable CAPS LOCK with double tap in Keyboard?

**Comment by MayuraVerma** [comment]
I have tried in iOS 15.0.188-alpha with EUROSIL
no sign of caps lock with double tap on shift key.

**Comment by MayuraVerma** [comment]
I tested full_caps_3620_3621.zip with iOS 15.0.188-alpha

CAPS lock turns off, after a key stroke
CAPS lock should stay in that state until turned off
if not there is not difference between CAPS lock and shift layer.

**Comment by mcdurdin** [comment]
@MakaraSok or @bharanidharanj, are you able to replicate the issues that @MayuraVerma has experienced? Can you work with Mayura to build a reproducible test case so we can resolve this before release?

**Comment by mcdurdin** [comment]
> I have tried in iOS 15.0.188-alpha with EUROSIL

The SIL Eurolatin keyboard has not yet been updated to support Caps Lock 😄

**Comment by mcdurdin** [comment]
> My earlier assumption was, web/ios/android picks up the NCAPS state from kmn.

Yeah, unfortunately given how the touch and physical keyboards diverge, we could not find a safe way to do this automatically. My preference is always to give the control to the keyboard designer, in any case -- but the trade-off is that we don't get this kind of behaviour "for free".

**Comment by mcdurdin** [comment]
> Once we enter CAPS layer, only why to get out of CAPS layer should be release the CAPS lock.

There are two ways, in our design, that the Caps Lock layer is released on touch:

1. When the user touches the Caps Lock key again (or another modifier key accessible on the Caps Lock layer).
2. When the user switches apps or input contexts. This matches the behaviour of the system keyboards -- Caps Lock state is not maintained long-term, unlike on a hardware keyboard.

**Pull request merged**